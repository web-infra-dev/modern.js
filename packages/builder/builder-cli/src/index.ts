export { defineConfig } from '@modern-js/builder/cli';
