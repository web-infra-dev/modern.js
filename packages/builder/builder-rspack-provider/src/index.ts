export { getRspackVersion } from './shared/rspackVersion';
export { builderRspackProvider } from './provider';
export type { BuilderRspackProvider } from './provider';

export type {
  // Config Types
  BuilderConfig,
  NormalizedConfig,

  // Hook Callback Types
  ModifyRspackConfigFn,

  // Plugin Types
  BuilderPluginAPI,

  // Rspack
  Rspack,
  RspackConfig,
} from './types';
