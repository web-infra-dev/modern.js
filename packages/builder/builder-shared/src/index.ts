export * from './applyDefaultBuilderOptions';
export * from './constants';
export * from './createHook';
export * from './pluginStore';
export * from './createContext';
export * from './utils';
export * from './fs';
export * from './generateMetaTags';
export * from './getBrowserslist';
export * from './getCssSupport';
export * from './logger';
export * from './mergeBuilderConfig';
export * from './onExitProcess';
export * from './pick';
export * from './regexp';
export * from './types';
export * from './config';
export * from './format';
export * from './devServer';
export * from './prodServer';
export * from './chain';
export * from './schema';
export * from './plugins';
export * from './openBrowser';
export * from './apply';
export * from './fallback';
export * from './getLoaderOptions';
export * from './svgo';
export * from './patch';
export * from './manifest';
export * from './css';
export * from './minimize';
export * from './core-js';
export * from './progress';
export * from './prettyTime';
