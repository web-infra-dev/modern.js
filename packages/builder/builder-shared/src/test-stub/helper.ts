import { Plugins } from '../types';

// eslint-disable-next-line @typescript-eslint/no-empty-function
const setup = () => {};

const genMockPlugin = (name: string) => () =>
  Promise.resolve({
    name,
    setup,
  });

export const mockBuilderPlugins: Plugins = {
  cleanOutput: genMockPlugin('builder-plugin-clean-output'),
  startUrl: genMockPlugin('builder-plugin-start-url'),
  fileSize: genMockPlugin('builder-plugin-file-size'),
  target: genMockPlugin('builder-plugin-target'),
  devtool: genMockPlugin('builder-plugin-devtool'),
  entry: genMockPlugin('builder-plugin-entry'),
  cache: genMockPlugin('builder-plugin-cache'),
  yaml: genMockPlugin('builder-plugin-yaml'),
  toml: genMockPlugin('builder-plugin-toml'),
  splitChunks: genMockPlugin('builder-plugin-split-chunks'),
  inlineChunk: genMockPlugin('builder-plugin-inline-chunk'),
  bundleAnalyzer: genMockPlugin('builder-plugin-bundle-analyzer'),
  assetsRetry: genMockPlugin('builder-plugin-assets-retry'),
  font: genMockPlugin('builder-plugin-font'),
  media: genMockPlugin('builder-plugin-media'),
  image: genMockPlugin('builder-plugin-image'),
  svg: genMockPlugin('builder-plugin-svg'),
  html: genMockPlugin('builder-plugin-html'),
  antd: genMockPlugin('antd'),
  arco: genMockPlugin('arco'),
  tsChecker: genMockPlugin('builder-plugin-ts-checker'),
  checkSyntax: genMockPlugin('builder-plugin-check-syntax'),
  rem: genMockPlugin('builder-plugin-rem'),
  wasm: genMockPlugin('builder-plugin-wasm'),
  moment: genMockPlugin('builder-plugin-moment'),
  nodeAddons: genMockPlugin('builder-plugin-node-addons'),
  externals: genMockPlugin('builder-plugin-externals'),
  sourceBuild: genMockPlugin('builder-plugin-source-build'),
  networkPerformance: genMockPlugin('builder-plugin-network-performance'),
  preloadOrPrefetch: genMockPlugin('builder-plugin-preload-or-prefetch'),
  performance: genMockPlugin('builder-plugin-performance'),
  define: genMockPlugin('builder-plugin-define'),
};
