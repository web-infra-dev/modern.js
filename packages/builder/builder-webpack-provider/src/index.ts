export { builderWebpackProvider } from './provider';
export type { BuilderWebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export type { WebpackBuildError } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  BuilderPlugin,
  BuilderPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  BuilderConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSLoaderOptions,
  CSSExtractOptions,
  LessLoaderOptions,
  SassLoaderOptions,
  HTMLPluginOptions,
  StyleLoaderOptions,
  AutoprefixerOptions,
  TerserPluginOptions,
  PostCSSLoaderOptions,
} from './types';
