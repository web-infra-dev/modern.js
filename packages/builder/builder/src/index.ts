export { createBuilder } from './createBuilder';
export { mergeBuilderConfig } from '@modern-js/builder-shared';

export { builderPluginSourceBuild } from './plugins/sourceBuild';

export type {
  BuilderMode,
  BuilderEntry,
  BuilderTarget,
  BuilderPlugin,
  BuilderContext,
  BuilderInstance,
  CreateBuilderOptions,
  InspectConfigOptions,

  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeCreateCompilerFn,
  OnDevCompileDoneFn,
  ModifyBuilderConfigFn,
} from '@modern-js/builder-shared';
