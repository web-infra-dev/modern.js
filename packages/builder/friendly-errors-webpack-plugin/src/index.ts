export * from './shared/types';
export * from './shared/utils';
export * from './plugin';
