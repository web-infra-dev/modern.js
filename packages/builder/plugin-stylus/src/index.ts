import type { BuilderPlugin } from '@modern-js/builder';
import { isUseCssSourceMap } from '@modern-js/builder-shared';
import type { BuilderPluginAPI } from '@modern-js/builder-webpack-provider';

type StylusOptions = {
  use?: string[];
  include?: string;
  import?: string;
  resolveURL?: boolean;
  lineNumbers?: boolean;
  hoistAtrules?: boolean;
};

type StylusLoaderOptions = {
  stylusOptions?: StylusOptions;
  sourceMap?: boolean;
};

export type PluginStylusOptions = StylusLoaderOptions;

export function builderPluginStylus(
  options?: PluginStylusOptions,
): BuilderPlugin<BuilderPluginAPI> {
  return {
    name: 'builder-plugin-stylus',

    async setup(api) {
      const { bundlerType } = api.context;
      api.modifyBundlerChain(async (chain, utils) => {
        const config = api.getNormalizedConfig();
        const { STYLUS_REGEX } = await import('@modern-js/builder-shared');
        const { applyOptionsChain } = await import('@modern-js/utils');

        const { merge: deepMerge } = await import('@modern-js/utils/lodash');

        const mergedOptions = applyOptionsChain<StylusLoaderOptions, undefined>(
          {
            sourceMap: isUseCssSourceMap(config),
          },
          options,
          undefined,
          deepMerge,
        );

        const rule = chain.module
          .rule(utils.CHAIN_ID.RULE.STYLUS)
          .test(STYLUS_REGEX);

        const { applyBaseCSSRule } = await import(
          `@modern-js/builder-${bundlerType}-provider/plugins/css`
        );
        await applyBaseCSSRule({
          rule,
          config: config as any,
          context: api.context,
          utils,
          importLoaders: 2,
        });

        rule
          .use(utils.CHAIN_ID.USE.STYLUS)
          .loader(require.resolve('stylus-loader'))
          .options(mergedOptions);
      });

      bundlerType === 'rspack' &&
        (api as any).modifyRspackConfig(async (rspackConfig: any) => {
          const { applyCSSModuleRule } = await import(
            '@modern-js/builder-rspack-provider/plugins/css'
          );
          const { STYLUS_REGEX } = await import('@modern-js/builder-shared');

          const config = api.getNormalizedConfig();

          const rules = rspackConfig.module?.rules;

          applyCSSModuleRule(rules, STYLUS_REGEX, config as any);
        });
    },
  };
}

/**
 * @deprecated Using builderPluginStylus instead.
 */
export const PluginStylus = builderPluginStylus;
