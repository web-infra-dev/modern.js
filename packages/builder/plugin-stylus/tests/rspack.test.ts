import { expect, describe, it } from 'vitest';
import { createBuilder } from '@rspack-builder/tests/helper';
import { builderPluginStylus } from '../src';

describe('plugins/stylus', () => {
  it('should add stylus loader config correctly', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginStylus() as any],
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should allow to configure stylus options', async () => {
    const builder = await createBuilder({
      plugins: [
        builderPluginStylus({
          stylusOptions: {
            lineNumbers: false,
          },
        }) as any,
      ],
    });
    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });
});
