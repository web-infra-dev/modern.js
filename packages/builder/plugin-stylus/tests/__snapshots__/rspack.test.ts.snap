// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/stylus > should add stylus loader config correctly 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "resolve": {
              "preferRelative": true,
            },
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/postcss-loader",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/stylus-loader/dist/cjs.js",
                "options": {
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "resolve": {
              "preferRelative": true,
            },
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/postcss-loader",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/stylus-loader/dist/cjs.js",
                "options": {
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.styl\\$/,
      },
    ],
  },
}
`;

exports[`plugins/stylus > should allow to configure stylus options 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "resolve": {
              "preferRelative": true,
            },
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/postcss-loader",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/stylus-loader/dist/cjs.js",
                "options": {
                  "sourceMap": true,
                  "stylusOptions": {
                    "lineNumbers": false,
                  },
                },
              },
            ],
          },
          {
            "resolve": {
              "preferRelative": true,
            },
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/postcss-loader",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/stylus-loader/dist/cjs.js",
                "options": {
                  "sourceMap": true,
                  "stylusOptions": {
                    "lineNumbers": false,
                  },
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.styl\\$/,
      },
    ],
  },
}
`;
